commit 8c34e2d63231d4bf4852bac8521883944d770fe3
Author: Jens Axboe <jens.axboe@oracle.com>
Date:   Tue Oct 17 19:43:22 2006 +0200

    [PATCH] Remove SUID when splicing into an inode
    
    Originally from Mark Fasheh <mark.fasheh@oracle.com>
    
    generic_file_splice_write() does not remove S_ISUID or S_ISGID. This is
    inconsistent with the way we generally write to files.
    
    Signed-off-by: Mark Fasheh <mark.fasheh@oracle.com>
    Signed-off-by: Jens Axboe <jens.axboe@oracle.com>

Backported to Debian's 2.6.18 by dann frazier <dannf@debian.org>

--- linux-source-2.6.18.orig/fs/splice.c	2008-10-03 17:14:24.000000000 -0600
+++ linux-source-2.6.18/fs/splice.c	2008-10-03 17:18:35.000000000 -0600
@@ -827,12 +827,21 @@
 			  loff_t *ppos, size_t len, unsigned int flags)
 {
 	struct address_space *mapping = out->f_mapping;
+	struct inode *inode = mapping->host;
 	ssize_t ret;
+	int err;
+
+	err = should_remove_suid(out->f_dentry);
+	if (unlikely(err)) {
+		mutex_lock(&inode->i_mutex);
+		err = __remove_suid(out->f_dentry, err);
+		mutex_unlock(&inode->i_mutex);
+		if (err)
+			return err;
+       }
 
 	ret = splice_from_pipe(pipe, out, ppos, len, flags, pipe_to_file);
 	if (ret > 0) {
-		struct inode *inode = mapping->host;
-
 		*ppos += ret;
 
 		/*
@@ -840,8 +849,6 @@
 		 * sync it.
 		 */
 		if (unlikely((out->f_flags & O_SYNC) || IS_SYNC(inode))) {
-			int err;
-
 			mutex_lock(&inode->i_mutex);
 			err = generic_osync_inode(inode, mapping,
 						  OSYNC_METADATA|OSYNC_DATA);
